add a simple tracing framework

zerbina opened the pull request:
## Summary
The goal is to provide the base for an event-based tracing framework
that can be incrementally improved/expanded upon. In its current form,
it's mainly meant for execution-time tracing.

While generally tailored towards being used as part of the compiler,
it's meant to be general enough to be used for both one-off temporary
instrumentation for use during debugging, and for building a persistent,
maybe even user-facing, compiler tracing facility. For example, it could
be expanded to also support the tracing done by `debugutils`.

This commit only adds the framework and a JSON serializer - both are not
used anywhere in the compiler yet.

## Details
The `tracer` module implements the data types and routines that make up
the framework.

The `trace_dump` module implements a serializer to the "Chrome Trace
Event" JSON Object format, which was chosen due to it being
simple-yet-powerful, with multiple visualizers already existing for it.

saem commented on compiler/utils/trace_dump.nim:
```diff
@@ -0,0 +1,134 @@
+    rawField(ph, ph); next()   # the event phase
+    floatField(ts, ts); next() # the timestamp
+
+    # while not it's not required by the specification for "End" events to store
+    # the "name" field, some visualizers won't work if the they don't 
```
```suggestion
    # while it's not required by the specification for "End" events to store
    # the "name" field, some visualizers won't work without 
```

saem commented on compiler/utils/trace_dump.nim:
```diff
@@ -0,0 +1,134 @@
+      if e.begin: BeginPhase
+      else:       EndPhase
+
+    # make the timestamp a relative one and change the unit its measured in
```
```suggestion
    # make the timestamp a relative one and change the unit it's measured in
```

saem reviewed (approved):
Two doc items they're not blockers.

This is awesome, thank you for doing it so quickly. 🥰

zerbina committed:
add a simple tracing framework

## Summary
The goal is to provide the base for an event-based tracing framework
that can be incrementally improved/expanded upon. In its current form,
it's mainly meant for execution-time tracing.

While generally tailored towards being used as part of the compiler,
it's meant to be general enough to be used for both one-off temporary
instrumentation for use during debugging, and for building a persistent,
maybe even user-facing, compiler tracing facility. For example, it could
be expanded to also support the tracing done by `debugutils`.

This commit only adds the framework and a JSON serializer - both are not
used anywhere in the compiler yet.

## Details
The `tracer` module implements the data types and routines that make up
the framework.

The `trace_dump` module implements a serializer to the "Chrome Trace
Event" JSON Object format, which was chosen due to it being
simple-yet-powerful, with multiple visualizers already existing for it.

zerbina commented:
I applied the suggestions (thanks @saem) and also improved the documentation of the `tracer` module a bit.

bors r+
